File: books/cs/general-purpose-algorithms/introduction-to-algorithms/3rd-edition/chapter2/C/insertion_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>

#define DATASETSIZE 10

void insertion_sort(double* srcdata, int len);
void main(void){
    double srcdata[DATASETSIZE];
    int idx, jdx;

    srand(time(NULL));
    for( idx=0; idx<DATASETSIZE; idx++ ){
        srcdata[idx] = rand()*1.0/RAND_MAX;
    }
    for( idx=0; idx<DATASETSIZE; idx++ ){
        if(idx%10 == 0) printf("\n\n");
        printf(" %f, ",srcdata[idx]);
    }
    printf("\n");
    insertion_sort(srcdata,DATASETSIZE);
    for( idx=0; idx<DATASETSIZE; idx++ ){
        if(idx%10 == 0) printf("\n\n");
        printf(" %f, ",srcdata[idx]);
    }
    printf("\n");
}

void insertion_sort(double* srcdata,int len){
	int idx = 0,jdx = 0;
	double dValue = 0;

	for( idx=1;idx<len;idx++ ){
		dValue = srcdata[idx];
		jdx = idx - 1;
		while( (jdx >= 0) && (srcdata[jdx]>dValue) ){
			srcdata[jdx+1] = srcdata[jdx];
			jdx = jdx - 1;
		} 
		srcdata[jdx+1] = dValue;
	}
}

File: books/cs/general-purpose-algorithms/introduction-to-algorithms/3rd-edition/chapter2/Scala/insertion_sort.scala
import scala.util.Random

object QuickSort{
    def main(args: Array[String]):Unit = {
        val dataSetSize = 10
        var srcdata = util.Random.shuffle(100 to 999).take(dataSetSize).toArray;

        println(s"the source array, ${srcdata.toList}, the length is ${srcdata.length}")
    	insertion_sort(srcdata)
        println(s"the result, ${srcdata.toList}")
    }

    def insertion_sort(srcdata: Array[Int]):Unit = {
		var idx: Int = 0
		var jdx: Int = 0
		var iValue: Int = 0
		for( idx <- 1 to (srcdata.length-1) ){
			iValue = srcdata(idx)
			jdx = idx - 1
			while( jdx >= 0 && iValue < srcdata(jdx) ){
				println(s"(${idx},${jdx}),${srcdata(jdx+1)},${srcdata(jdx)}, ${iValue}")
				srcdata(jdx+1) = srcdata(jdx)
				jdx = jdx - 1
			}
			srcdata(jdx+1) = iValue
            println(s"jdx=${jdx},${srcdata.toList}")
		}
    }

}
